### PR TITLE
feat: added build-context support in ci build

### LINKED ISSUE #2703: Feature: Deployment Support for  Monorepo
### 🔖 Feature description

Current Deployment setup downloads the git repo and tries to build docker image from root path.

Repo Structrure 
- .gitignore
- service1/Dockerfile
- service1/scr/*
- service1/package.json
- service2/Dockerfile
- service2/scr/*
- service2/package.json

example of Dockerfile
path: <root>/service1/Dockerfile

```
FROM keymetrics/pm2:12-alpine

ENV HOME=/home/node
USER root
WORKDIR $HOME/app
COPY package.json $HOME/tmp/
COPY package-lock.json $HOME/tmp/
COPY .npmrc $HOME/tmp/
RUN cd $HOME/tmp/ && npm install
COPY src $HOME/app/src
RUN cd $HOME/app && npm run build
COPY pm2.json ./

EXPOSE 9026
```

### 🎤 Pitch

While image build, deployment script should have an option to moved to a directory (sub path) in the repo and then perform the `docker build ` command.

### ✌️ Solution

While image build, deployment script should have an option to moved to a directory (sub path) in the repo and then perform the `docker build ` command.

### 🔄️ Alternative

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by gireesh-devtron**
# Description

Fixes #2703 #2070

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [x] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/pipeline/CiService.go [hunk 1]
```diff
@@ -428,12 +429,18 @@ func (impl *CiServiceImpl) buildWfRequestForCiPipeline(pipeline *pipelineConfig.
...
 	oldArgs := ciTemplate.Args
-	ciBuildConfigBean, err = bean2.OverrideCiBuildConfig(dockerfilePath, oldArgs, ciLevelArgs, ciTemplate.DockerBuildOptions, ciTemplate.TargetPlatform, ciBuildConfigBean)
+	ciBuildConfigBean, err = bean2.OverrideCiBuildConfig(dockerfilePath, oldArgs, ciLevelArgs, ciTemplate.DockerBuildOptions, ciTemplate.TargetPlatform, ciTemplate.BuildContext, ciBuildConfigBean)
 	if err != nil {
 		impl.Logger.Errorw("error occurred while overriding ci build config", "oldArgs", oldArgs, "ciLevelArgs", ciLevelArgs, "error", err)
 		return nil, errors.New("error while parsing ci build config")
 	}
+	buildContextCheckoutPath, err := impl.ciPipelineMaterialRepository.GetCheckoutPath(ciBuildConfigBean.BuildContextGitMaterialId)
+	if err != nil {
+		impl.Logger.Errorw("error occurred while getting checkout path from git material", "gitmaterial-id", ciBuildConfigBean.BuildContextGitMaterialId, "error", err)
```
Comment (kartik-579) [inline pkg/pipeline/CiService.go:439]: use camelcase got gitmaterial-id

File: pkg/pipeline/CiTemplateService.go [hunk 2]
```diff
@@ -140,12 +141,13 @@ func (impl CiTemplateServiceImpl) extractBuildConfigBean(templateOverride *pipel
 		return nil, err
 	}
 	if ciBuildConfigBean == nil {
-		ciBuildConfigBean, err = bean.OverrideCiBuildConfig(templateOverride.DockerfilePath, "", "", "", "", nil)
+		ciBuildConfigBean, err = bean.OverrideCiBuildConfig(templateOverride.DockerfilePath, "", "", "", "", ".", nil)
```
Comment (kartik-579) [inline pkg/pipeline/CiTemplateService.go:144]: use const for "."

**Review by kartik-579 — CHANGES_REQUESTED** [review]

**Review by kartik-579 — APPROVED** [review]

File: api/appbean/AppDetail.go [hunk 3]
```diff
@@ -58,6 +58,7 @@ type DockerBuildConfig struct {
 	Args                   map[string]string `json:"args,omitempty"`
 	TargetPlatform         string            `json:"targetPlatform"`
 	DockerBuildOptions     map[string]string `json:"dockerBuildOptions,omitempty"`
+	BuildContext           string            `json:"buildContext"`
```
[the remark this answers is not in the document]
Comment (gireesh-devtron) [reply api/appbean/AppDetail.go]: done

File: pkg/pipeline/CiTemplateService.go (hunk 2 above)
Comment (gireesh-devtron) [reply pkg/pipeline/CiTemplateService.go]: done

**Review by kartik-579 — APPROVED** [review]

**Review by kartik-579 — APPROVED** [review]

**Pull request merged**